sw/device/lib/dif/dif_gpio.c

```diff
@@ -0,0 +1,358 @@
+ * upper half of DIRECT_OUT.
+ *
+ * @param gpio GPIO instance.
+ * @param lower_offset Offset of the masked access register that corresponds to
+ * the lower half of the actual register.
+ * @param upper_offset Offset of the masked access register that corresponds to
+ * the upper half of the actual register.
+ * @param mask Mask that identifies the bits to write to.
+ * @param val Value to write.
+ */
+static void gpio_masked_write(const dif_gpio_t *gpio, uint32_t lower_offset,
+                              uint32_t upper_offset, uint32_t mask,
+                              uint32_t val) {
+  const uint32_t mask_lower_half = mask & 0x0000FFFFU;
+  const uint32_t mask_upper_half = mask & 0xFFFF0000U;
+  if (mask_lower_half != 0) {
+    mmio_region_write32(gpio->base_addr, lower_offset,
+                        (mask_lower_half << 16) | (val & 0x0000FFFFU));
+  }
+  if (mask_upper_half != 0) {
+    mmio_region_write32(gpio->base_addr, upper_offset,
+                        mask_upper_half | ((val & 0xFFFF0000U) >> 16));
+  }
+}
+
+/**
+ * Perform a masked write to a single bit of a GPIO register.
+ *
+ * GPIO device provides masked (thread-safe) writes to some of its registers.
+ * This allows software to modify half of the bits at a time without requiring
+ * a read-modify-write. This function is guaranteed to perform only one write
+ * since it never needs to access both halves of a register.
+ *
+ * See also |gpio_masked_write()|.
+ *
+ * @param gpio GPIO instance.
+ * @param lower_offset Offset of the masked access register that corresponds to
+ * the lower half of the actual register.
+ * @param upper_offset Offset of the masked access register that corresponds to
+ * the upper half of the actual register.
+ * @param index Zero-based index of the bit to write to.
+ * @param val Value to write.
+ */
+static void gpio_masked_bit_write(const dif_gpio_t *gpio, uint32_t lower_offset,
+                                  uint32_t upper_offset, uint32_t index,
+                                  bool val) {
+  // Write to lower_offset if the bit is in the lower half, write to
+  // upper_offset otherwise.
+  const ptrdiff_t offset = (index < 16) ? lower_offset : upper_offset;
+  // Since masked access writes to half of a register, index mod 16 gives the
+  // index of the bit in the half-word mask.
+  const uint32_t mask = index_to_mask(index % 16);
+  mmio_region_write32(gpio->base_addr, offset,
+                      (mask << 16) | (val ? mask : 0U));
+}
+
+dif_gpio_result_t dif_gpio_init(const dif_gpio_config_t *config,
+                                dif_gpio_t *gpio) {
+  if (config == NULL || gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  // Save internal state in the given |dif_gpio_t| instance.
+  gpio->base_addr = config->base_addr;
+  // Reset the GPIO device at the given |base_addr|.
+  dif_gpio_reset(gpio);
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_reset(const dif_gpio_t *gpio) {
```

Comment:
(Please also see my comment in your PR) I would prefer a device instance over a base address (like any other public DIF library function).

My answer to your question about the bootrom would be "it depends". Devices should already be in a _reasonable_ default state after power up and coming out of reset. And, as the name implies, I would expect a reset function to reset a device to this initial state. So, I'm not sure if there is a compelling reason to call reset functions in the bootrom. I think, it should be the responsibility of the code in the flash to configure the devices that it will use, unless I'm missing a case where the initialization of a device requires something only available in bootrom. Of course, if the bootrom needs to use a device, e.g. for debug purposes, then it should initialize and use it as usual. In that case, we just need to see if we need any special handling as the control passes from bootrom to flash.

Reply:
Sorry, I have missed out this comment, I didn't mean to move the conversation to the UART PR #1483 . If you could have a look at the latest comment on that PR and I am happy to discuss further here, or on that PR.

Reply:
No worries! See you at that PR :)